kernel/mem_domain.c

```diff
@@ -0,0 +1,294 @@
+#include <kernel.h>
+#include <kernel_structs.h>
+#if defined(CONFIG_ARM_CORE_MPU)
+#include <arch/arm/cortex_m/mpu/arm_core_mpu.h>
+#endif
+
+
+static u8_t max_partitions;
+
+
+static inline u8_t arch_get_max_partitions(void)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	return arm_core_mpu_get_max_domain_partition_regions();
+#else
+	return CONFIG_MAX_DOMAIN_PARTITIONS;
+#endif
+}
+
+static inline void arch_configure_mem_domain(struct k_thread *thread)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	configure_mpu_mem_domain(thread);
+#endif
+}
+
+static inline void arch_configure_mem_partition(struct k_thread *thread,
+						u32_t part_index)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	configure_mpu_mem_partition(thread, part_index);
+#endif
+}
+
+static inline void arch_disable_mem_partition(struct k_thread *thread,
+					      u32_t part_index)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	disable_mpu_mem_partition(thread, part_index);
+#endif
+}
+
+static void arch_disable_all_mem_partitions(struct k_thread *thread)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	disable_mpu_all_mem_partitions(thread);
+#endif
+}
+
+
+int k_mem_domain_init(struct k_mem_domain *domain, const char *name,
+		      u32_t num_parts, struct k_mem_partition *parts[])
+{
+	unsigned int key;
+
+	if (!domain || !name || (num_parts && !parts)) {
+		return -EINVAL;
+	}
+
+	if (num_parts > max_partitions) {
+		return -ENOMEM;
+	}
+
+	key = irq_lock();
+
+	domain->name = name;
+	domain->num_partitions = num_parts;
+	memset(domain->partitions, 0, sizeof(domain->partitions));
+
+	if (num_parts) {
+		u32_t i;
+
+		for (i = 0; i < num_parts; i++) {
+			if (!parts[i]) {
+				irq_unlock(key);
+				return -EINVAL;
+			}
+			domain->partitions[i] = *parts[i];
+		}
+	}
+
+	sys_dlist_init(&domain->mem_domain_q);
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+int k_mem_domain_destroy(struct k_mem_domain *domain)
+{
+	unsigned int key;
+	sys_dnode_t *node, *next_node;
+
+	if (!domain) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
+		struct k_thread *thread =
+			CONTAINER_OF(node, struct k_thread, mem_domain_info);
+
+		/*
+		 * Call arch specific API to disable thread's memory
+		 * partitions.
+		 */
+		arch_disable_all_mem_partitions(thread);
+
+		sys_dlist_remove(&thread->mem_domain_info.mem_domain_q_node);
+		thread->mem_domain_info.mem_domain = NULL;
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+int k_mem_domain_add_partition(struct k_mem_domain *domain,
+			       struct k_mem_partition *part)
+{
+	int p_idx;
+	unsigned int key;
+	sys_dnode_t *node, *next_node;
+
+	if (!domain || !part) {
+		return -EINVAL;
+	}
+
+	if (part->start + part->size <= part->start) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	for (p_idx = 0; p_idx < max_partitions; p_idx++) {
+		/* A zero-sized partition denotes it's a free partition */
+		if (domain->partitions[p_idx].size == 0) {
+			break;
+		}
+	}
+
+	/* No free partition */
+	if (p_idx == max_partitions) {
```

Comment:
I think this should be an __ASSERT() too

Reply:
revised.